test/e2e/foursquare.e2e.test.js

```diff
@@ -0,0 +1,116 @@
+
+    // At least one result row
+    const rows = table.locator('tbody tr');
+    const rowCount = await rows.count();
+    expect(rowCount).toBeGreaterThan(0);
+
+    // Validate first row structure and formats
+    const firstRowCells = rows.first().locator('td');
+    await expect(firstRowCells).toHaveCount(5);
+
+    // Icon cell: either an <img> category icon or literal "N/A"
+    const iconImgCount = await firstRowCells.nth(0).locator('img').count();
+    if (iconImgCount > 0) {
+      const icon = firstRowCells.nth(0).locator('img');
+      await expect(icon).toHaveAttribute('src', /https?:\/\//);
+      await expect(icon).toHaveAttribute('alt', /\w+/);
+      const w = await icon.getAttribute('width');
+      const h = await icon.getAttribute('height');
+      expect(['32', '36', '40', '48', '64']).toContain(w); // size can vary in future; allow a few common values
+      expect(h).toBe(w);
+    } else {
+      const text = (await firstRowCells.nth(0).textContent()).trim();
+      expect(text).toBe('N/A');
+    }
+
+    // Name cell: non-empty
+    const venueName = (await firstRowCells.nth(1).textContent()).trim();
+    expect(venueName.length).toBeGreaterThan(0);
+
+    // Category cell: either non-empty string or 'N/A'
+    const categoryText = (await firstRowCells.nth(2).textContent()).trim();
+    expect(categoryText.length > 0 || categoryText === 'N/A').toBeTruthy();
+
+    // Address cell: string (may be 'N/A' depending on data)
+    const addrText = (await firstRowCells.nth(3).textContent()).trim();
+    expect(addrText.length > 0).toBeTruthy();
+
+    // Distance cell: numeric
+    const distanceText = (await firstRowCells.nth(4).textContent()).trim();
+    expect(distanceText).toMatch(/^\d+$/);
+  });
+
+  test('should render Venue Details with name, category, and coordinates', async () => {
+    // Section header
+    await expect(sharedPage.locator('h3.text-primary', { hasText: 'Venue Details' })).toBeVisible();
+
+    // The details paragraph contains <i><u>name</u></i>, optional category, and location + lat/long
+    const detailsPara = sharedPage.locator('h3.text-primary:has-text("Venue Details") + p');
+    await expect(detailsPara).toBeVisible();
+
+    // Name element
+    const nameElement = detailsPara.locator('i u');
+    await expect(nameElement).toBeVisible();
+    const detailName = (await nameElement.textContent()).trim();
+    expect(detailName.length).toBeGreaterThan(0);
+
+    // Should include the phrases that come from API-driven template
+    const detailsText = (await detailsPara.textContent()).toLowerCase();
+    expect(detailsText).toContain('located at');
+    expect(detailsText).toContain('longitude');
+    expect(detailsText).toContain('latitude');
+
+    // If category phrase is present, it should read like: "is a <Category>"
+    if (detailsText.includes(' is a ')) {
+      const hasIsAPhrase = /\bis a\b/.test(detailsText);
+      expect(hasIsAPhrase).toBe(true);
+    }
+  });
```

Comment:
Also check that we have  list of "Related venues or businesses to Pike Place Market" that has 10 or more venues.